### DISCUSSION TITLE
Map object types ?

### DISCUSSION THREAD

**Discussion by jrhoward** [savonet/liquidsoap#3108] [General]
I am trying iterate over a list ( list-a )  which is ordered and in the context of  iterator object retrieve the matching object from another list (list-b) to create another list (list-c) which is ordered the same as list -a. It would be much easier if I could have an Map type object for list-b and retrieve by key but I cannot find any references to map objects. I have to iterate over list-b for every item in list-a which is very inefficient.  

It's likely it  just because I am not well versed in functional programming. I am wondering if I am missing something and if there is a functional pattern to achieve  the above efficiently.

**Comment by jrhoward** [comment]
ok  found association lists in liquidsoap book.

**Reply by vitoyucepi** [reply]
Hi @jrhoward,
Do you still have a problem?

**Comment by jrhoward** [comment]
Hey @vitoyucepi 
Yes, association list is not what I thought it was.
I have a solution but it seems very inefficient running list.iter inside list.iter

**Reply by vitoyucepi** [reply]
Did you find [`list.assoc.mem`](https://www.liquidsoap.info/doc-2.1.4/reference.html#list.assoc.mem) and [`list.assoc`](https://www.liquidsoap.info/doc-2.1.4/reference.html#list.assoc)?
`list.assoc.mem` checks if a key is among the keys of the associated list.
`list.assoc` gets value for the given key.

**Comment by jrhoward** [comment]
I did investigate that. The example in the book  uses:

```
m = [("artist", "Frank Sinatra"), ("title", "Fly me to the moon")]
list.assoc("title", m)
```

I'm looking for something more like:

```
m = [ { filename="...", title="...", artist="..." }, { filename="...", title="...", artist="..." } , ... ]

list.iter(  fun(filename) -> get object from m where m.filename == filename, playlist.remaining_files() )


```
where the new list is ordered like the remaining songs in the playlist
With a map I could do this easily because I could associate the value of `filename` with a key.

**Reply by vitoyucepi** [reply]
Are you trying to implement something like the [`playlog`](https://www.liquidsoap.info/doc-2.1.4/reference.html#playlog)?
Do you have any prior experience in python or js?

**Reply by toots** [reply]
Should this be a fold?

```liquidsoap
m = [ { filename="...", title="...", artist="..." }, { filename="...", title="...", artist="..." } , ... ]

filtered_m = list.fold(fun (filtered_m, el) ->
  if (m.filename == "some value") then
    [...filtered_m, el]
  else
    filtered_m
  end, [], m)
```

**Reply by jrhoward** [reply]
Thanks toots I have not investigated fold. Will have a look at that. 

vitoyucepi, yes I have a few programming languages under my belt but only want to do that if it's not achievable in liquidsoap.
If I were to it implement outside of liquidsoap I would probably use golang.

I appreciate your help !!

**Reply by jrhoward** [reply]
regarding the `playlog` question, I am rendering what is left in a playlist in order using `remaining_files()` but accessing the files to extract metadata is quite expensive.

I run a function `on_reload` in a thread for the playlist to create a list of song objects. I am reusing this with `remaining_files` to access song information rather than reloading metadata from the file system.

this code works but there is an iterator inside the iterator
The root iterator needs to be remaining_files()  to preserve order:

```
// song_list is the list containing all songs generated on_reload

songs = ref([])

def add_song( filename )
    list.iter( fun( song )-> ignore( string.contains(suffix=song.filename, filename )?songs:=list.add( song, !songs ): null() ),!song_list )
end

list.iter( add_song , default_playlist.remaining_files()  )

```

**Reply by vitoyucepi** [reply]
> yes I have a few programming languages under my belt but only want to do that if it's not achievable in liquidsoap

It will be easier for me to translate a solution from another programming language than from a natural language.
For example
```python
m = [{"filename": "/tmp/test/3.mp3", "title": "3", "artist": "3"},
     {"filename": "/tmp/test/2.mp3", "title": "2", "artist": "2"}]

playlist_remaining_files = ["/tmp/test/1.mp3", "/tmp/test/2.mp3", "/tmp/test/3.mp3", "/tmp/test/4.mp3"]

for filename in playlist_remaining_files:
    for item in m:
        if item["filename"] == filename:
            print(item)

```
will translate into
```ruby
m = [{filename = "/tmp/test/3.mp3", title = "3", artist = "3"},
     {filename = "/tmp/test/2.mp3", title = "2", artist = "2"}]
p = playlist("/tmp/test", mode="normal")

for filename = list.iterator(p.remaining_files()) do
  for item = list.iterator(m) do
    if item.filename == filename then
      print(item)
    end
  end
end
```

For me, it's not clear what are you planning to do.
Maybe it'll be better to store `m` as
```ruby
m = [
  ("/tmp/test/3.mp3", {filename = "/tmp/test/3.mp3", title = "3", artist = "3"}),
  ("/tmp/test/2.mp3", {filename = "/tmp/test/2.mp3", title = "2", artist = "2"}),
]
```
And then
```ruby
if list.assoc.mem(filename, m) then
  list.assoc(filename, m)
end
```

**Reply by jrhoward** [reply]
thanks @vitoyucepi, that will work.
This structure
```
m = [
  ("/tmp/test/3.mp3", {filename = "/tmp/test/3.mp3", title = "3", artist = "3"}),
  ("/tmp/test/2.mp3", {filename = "/tmp/test/2.mp3", title = "2", artist = "2"}),
]
```
I was under the impression that  `list.assoc` only worked with (string,string)
